File: mlx-sys/src/fast.cpp
#include "mlx-cxx/fast.hpp"
#include "mlx-cxx/utils.hpp"

namespace mlx_cxx::fast {
        std::unique_ptr<mlx::core::array> rope(
        const mlx::core::array& x,
        int dims,
        bool traditional,
        float base,
        float scale,
        int offset,
        mlx_cxx::StreamOrDevice s)
    {
        auto arr = mlx::core::fast::rope(
            x,
            dims,
            traditional,
            base,
            scale,
            offset,
            s.to_variant());
        return std::make_unique<mlx::core::array>(arr);
    }

    std::unique_ptr<mlx::core::array> scaled_dot_product_attention(
        const mlx::core::array & queries,
        const mlx::core::array & keys,
        const mlx::core::array & values,
        const float scale,
        const mlx_cxx::OptionalArray & mask,
        mlx_cxx::StreamOrDevice s
    ) {
        auto std_mask = mlx_cxx::to_std_optional(mask);

        auto arr = mlx::core::fast::scaled_dot_product_attention(
            queries,
            keys,
            values,
            scale,
            std_mask,
            s.to_variant());

        return std::make_unique<mlx::core::array>(arr);
    }
}

File: mlx-sys/src/fast.rs
#[cxx::bridge]
pub mod ffi {
    unsafe extern "C++" {
        include!("mlx-cxx/fast.hpp");

        #[namespace = "mlx_cxx"]
        type OptionalArray = crate::ops::ffi::OptionalArray;

        #[namespace = "mlx::core"]
        type array = crate::array::ffi::array;

        #[namespace = "mlx_cxx"]
        type StreamOrDevice = crate::utils::StreamOrDevice;

        #[namespace = "mlx_cxx::fast"]
        fn rope(
            x: &array,
            dims: i32,
            traditional: bool,
            base: f32,
            scale: f32,
            offset: i32,
            s: StreamOrDevice,
        ) -> Result<UniquePtr<array>>;

        #[namespace = "mlx_cxx::fast"]
        fn scaled_dot_product_attention(
            queries: &array,
            keys: &array,
            values: &array,
            scale: f32,
            mask: &OptionalArray,
            s: StreamOrDevice,
        ) -> Result<UniquePtr<array>>;

        #[namespace = "mlx_cxx::fast"]
        fn rms_norm(
            x: &array,
            weight: &array,
            eps: f32,
            s: StreamOrDevice,
        ) -> Result<UniquePtr<array>>;

        #[namespace = "mlx_cxx::fast"]
        fn layer_norm(
            x: &array,
            weight: &OptionalArray,
            bias: &OptionalArray,
            eps: f32,
            s: StreamOrDevice,
        ) -> Result<UniquePtr<array>>;
    }
}
